src/rgw/setup/rgw.py

```diff
@@ -187,36 +187,78 @@ def cleanup(conf: MappedConf, pre_factory: bool = False):
+        Conf.save(config_index)
+
+    @staticmethod
+    def _remove_rgw_property(conf: MappedConf, config_index:str, config_key: str):
+        """Remove specific key from rgw config file."""
+        for rgw_config_key, confstore_key in const.SVC_CONFIG_DICT.items():
+            if confstore_key == config_key:
+                Conf.delete(config_index, f'{const.CLIENT_SECTION}>{rgw_config_key}')
+        Conf.save(config_index)
+
+    @staticmethod
+    def upgrade(conf: MappedConf, changeset_path: str):
         """Perform upgrade steps."""
         Log.info('Upgrade phase started.')
         conf_dir = Rgw._get_rgw_config_dir(conf)
         svc_conf_file = Rgw._get_rgw_config_path(conf)
+
+        # Load changeset file
+        changeset_index="rgw_changeset_index"
+        if not os.path.exists(changeset_path):
+             raise SetupError(errno.EINVAL, f'Changeset file is missing for Upgrade.')
+        Rgw._load_rgw_config(changeset_index, changeset_path)
+
+        # Get all changed keys from changeset file.
+        changeset_all_keys=Conf.get_keys(changeset_index)
+
         # Load deployed rgw config and take a backup.
         Rgw._load_rgw_config(Rgw._conf_idx, const.CONFSTORE_FILE_HANDLER + svc_conf_file)
         deployed_version = conf.get(const.VERSION_KEY)
         conf_bkp_file = os.path.join(conf_dir, const.RGW_CONF_FILE + f'.{deployed_version}')
+
         try:
+            # Cleaning up failure config of previous upgrade.
             if os.path.exists(conf_bkp_file):
-                # Cleaning up failure config of previous upgrade.
-                os.rename(conf_bkp_file, svc_conf_file)
-            else:
-                os.rename(svc_conf_file, conf_bkp_file)
-            # Generate new updated config.
-            tmpl_idx = f'{const.COMPONENT_NAME}_conf_tmpl'  # e.g. rgw_conf_tmpl
-            Rgw._load_rgw_config(tmpl_idx, const.CONFSTORE_FILE_HANDLER + const.CONF_TMPL)
-            Conf.copy(tmpl_idx, Rgw._conf_idx)
-            Conf.save(Rgw._conf_idx)
-            Rgw._create_svc_config(conf)
-            # TODO: separate out user creation and update_hax_endpoint logic from
-            # _update_hax_endpoint_and_create_admin() fun and add update_hax_endpoint
-            # in _create_svc_config and remove below lines 200-201.
-            data_nodes = Rgw._get_data_nodes(conf)
-            Rgw._update_hax_endpoint(conf, data_nodes[0])
+                os.remove(conf_bkp_file)
+
+            # create backup of existing config file
+            os.copy(svc_conf_file, conf_bkp_file)
+
+            # Handle svc key & Gconf key mapping
+            for key in changeset_all_keys:
+                if key.stratswith('new') :
+                    # Handle addition of new key
+                    value = Conf.get(changeset_index, key)
+                    key =key.split('new>')[1]
+                    Rgw._update_rgw_property(conf, Rgw._conf_idx, key, new_val)
```

Comment:
this function doesnt help with addition of new config as i see

Reply:
yes . this will check if corresponding rgw config is present in integration code (const.py mapping directory) If it present then it will add or else it will skip this addition entry.

Reply:
yes, i understand now